### PR TITLE
Add tls to ftp

### PR DESCRIPTION
**Pull request by bartvaneswhiffle**
See #1580.

It is basically a copy of the FTP class in fsspec, with a modification to use FTP_TLS and one addtional change.

Please have a look at [this](https://github.com/fsspec/filesystem_spec/pull/1581/commits/1937cc8ccbb0757141feb7156b2c58c0126e8cfe#diff-c232dd1a6b61794c2f4158c99b714b6c89d021d2d84af2bcd784c2a74f86f236R83).

### CODE REVIEW

File: fsspec/implementations/ftp_tls.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+        if sys.version_info >= (3, 9):
+            self.ftp = FTP_TLS(timeout=self.timeout, encoding=self.encoding)
+        elif self.encoding:
+            warnings.warn("`encoding` not supported for python<3.9, ignoring")
+            self.ftp = FTP_TLS(timeout=self.timeout)
+        else:
+            self.ftp = FTP_TLS(timeout=self.timeout)
+        self.ftp.connect(self.host, self.port)
+        self.ftp.login(*self.cred)
+        self.ftp.prot_p()
```
Comment (bartvaneswhiffle) [inline fsspec/implementations/ftp_tls.py:83]: Please take a look at this.

**Comment by martindurant** [comment]
Thanks for contributing!

Before going further, do you think it's possible to provide the extra features as options on the original FTP filesystem rather than create a new variant that is largely the same?

**Comment by bartvaneswhiffle** [comment]
I've removed the the `ftp_tls` class and integrated it with the `ftp` class. Two tests are currently still failing: the test with caching (test_complex) and the test with transaction (test_transaction), not sure how to fix those?

File: fsspec/implementations/tests/test_ftp_tls.py
Comment (martindurant) [inline fsspec/implementations/tests/test_ftp_tls.py:1]: Is this an exact copy of he insecure tests? You can use pytest.mark.parametrize to run each of the existing tests, once with secure=False and once with True.

Comment (bartvaneswhiffle) [reply fsspec/implementations/tests/test_ftp_tls.py:1]: This is not possible since it will need a different fixture.

**Comment by martindurant** [comment]
Ping: is this PR still progressing?

**Comment by bartvaneswhiffle** [comment]
Thanks for the reminder, so far I've not had the time to further implement the tests.

**Comment by bartvaneswhiffle** [comment]
I've found some time to update the tests. The pyftpdlib server has been replaced with script that starts a pyftpdlib server that accepts TLS. Furtermore, docstrings and codestyle improvements have been made.

@martindurant: Do you understand why the [CI / downstream fails](https://github.com/fsspec/filesystem_spec/actions/runs/10301202703/job/28512219710?pr=1581)?

**Comment by bartvaneswhiffle** [comment]
> > Do you understand why the [CI / downstream fails](https://github.com/fsspec/filesystem_spec/actions/runs/10301202703/job/28512219710?pr=1581)
> 
> Please merge from master and try again

All checks have passed :)

**Pull request merged**